### PR TITLE
Issue #469 rescaling of plot graphs

### LINKED ISSUE #469: Plot graphs do not rescale when combined with plots from other spectrogram scales
## Actual behaviour:

In a generic recogniser one can use different spectrogram settings for each profile. If the window sizes are different then the diagnostic plots produced are rendered with different amounts of data and produce a not so useful image.

## How to reproduce this bug:

1. Create a generic recogniser with two profiles
2. Ensure each profile uses a different spectrogram frame size
3. View the output debug spectrogram

## Additional Details

K-P-THINNING_20130209_062400_0-0 5min__Image_not_cheating

You can see one of the plots is much longer than the rest of the content in the image.

### PR DESCRIPTION
**Pull request by towsey**
# Rescaling of plots 

Rescales plots before rendering them in generic recognisers. Fixes #469 

## Changes
Wrote a unit test to demonstrate that the rescaling of plots is working.

## Issues
None. Except that also change the FFT window to Hanning in MultiRecognizer.cs.
That may or maynot cause a change somewhere.

## Final Checklist

-   [Anthony ] Assign reviewers if you have permission
-   [ ] Assign labels if you have permission
-   [ ] Link issues related to PR
-   [ ] Link any PRs or issues blocking this PR from being merged
-   [ ] Remove/Reduce warnings from edited files
-   [ Yes] Unit tests have been added for changes
-   [ ] Ensure CI build is passing

### CODE REVIEW

**Comment by atruskie** [comment]
Hi @towsey. Good tests but the issue was with the generic recogniser and not the multi recogniser.

You probably just need to add the array scaling call in the generic recogniser class.

But the way, this is what the `imageWidth` argument is used for (your comments frequently indicate you don't know what it is for). It's used to define a fixed width images should be when rendered. Not applicable here but I was reminded by the subject.

**Comment by towsey** [comment]
I have included a loop to rescale plots to fit the accompanying spectrogram.

File: tests/Acoustics.Test/AnalysisPrograms/Recognizers/GenericRecognizerTests.cs [hunk 1]
```diff
@@ -310,10 +310,9 @@ public void TestWhistleAlgorithm()
 
             var results = recognizer.Recognize(recording, config, 100.Seconds(), null, this.TestOutputDirectory, null);
 
-            // add next two lines because cannot find spectrogram included with the test results.
             // Used for debugging only
-            var image = SpectrogramTools.GetSonogramPlusCharts(results.Sonogram, results.NewEvents, results.Plots, null);
-            image.SaveAsPng("C:/temp/image.png");
+            var image1 = SpectrogramTools.GetSonogramPlusCharts(results.Sonogram, results.NewEvents, results.Plots, null);
+            image1.Save(this.TestOutputDirectory + "\\image1.png");
```
Comment (atruskie) [inline tests/Acoustics.Test/AnalysisPrograms/Recognizers/GenericRecognizerTests.cs:315]: You cannot use backslashes - our tests run on non-windows paltforms.

Use the `this.TestOutputDirectory.CombineFile("WhistleSpectrogram.png")`

File: tests/Acoustics.Test/TowseyLibrary/DataToolsTests.cs [hunk 2]
```diff
@@ -38,5 +41,85 @@ public void TestConcatenateVectorsOverload()
...
+            var imageList1 = new List<Image<Rgb24>>();
+            foreach (var plot in plots)
+            {
+                var image = plot.DrawAnnotatedPlot(plotHeight);
+                imageList1.Add(image);
+            }
+
+            // create the image for visual confirmation
+            var concatImage1 = ImageTools.CombineImagesVertically(imageList1);
+            concatImage1.Save(this.TestOutputDirectory + "ConcatImage1.png");
```
Comment (atruskie) [inline tests/Acoustics.Test/TowseyLibrary/DataToolsTests.cs:99]: Do not join paths with the `+` operator.

Use the `SaveTestOutput` function

This comment applies to all usages in this class.

**Review by atruskie — CHANGES_REQUESTED** [review]

**Comment by towsey** [comment]
Very soryy but unfortunately, the work involved in fixing this aparnetly small Plot scaling problem got significantly more difficult because I used the OscillationEvent class in the unit test. It turned out that the functionality for much of Oscillation events was not working. So this latest commit is mostly about fixing oscillation events.

File: tests/Acoustics.Test/AnalysisPrograms/Recognizers/GenericRecognizerTests.cs [hunk 3]
```diff
@@ -1204,5 +1208,83 @@ public void TestMultipleAlgorithms()
...
+                            MaxHertz = 7200,
+                            MinHertz = 4800,
+                            BgNoiseThreshold = 0.0,
+                            BottomHertzBuffer = 1000,
+                            TopHertzBuffer = 500,
+                            DecibelThresholds = new double?[] { 3.0 },
+                        }
+                    },
+                    {
+                        "LowerBandDTMF_z",
```
Comment (atruskie) [inline tests/Acoustics.Test/AnalysisPrograms/Recognizers/GenericRecognizerTests.cs:1233]: the name of the test method and test profile do not match up.

DTMF tones are harmonic not oscillations. 

Probably not wrong but it is confusing

**Review by atruskie — APPROVED** [review]
I've suggested one change. When you're ready let me know and I'll merge this in

**Pull request merged**